Update endpoint to return 404 after isOptimistic check

Muzry opened the pull request:
<!-- Thanks for sending a PR! Before submitting:

1. If this is your first PR, check out our contribution guide here https://docs.prylabs.network/docs/contribute/contribution-guidelines
   You will then need to sign our Contributor License Agreement (CLA), which will show up as a comment from a bot in this pull request after you open it. We cannot review code without a signed CLA.
2. Please file an associated tracking issue if this pull request is non-trivial and requires context for our team to understand. All features and most bug fixes should have
   an associated issue with a design discussed and decided upon. Small bug
   fixes and documentation improvements don't need issues.
3. New features and bug fixes must have tests. Documentation may need to
   be updated. If you're unsure what to update, send the PR, and we'll discuss
   in review.
4. Note that PRs updating dependencies and new Go versions are not accepted.
   Please file an issue instead.
5. A changelog entry is required for user facing issues.
-->

**What type of PR is this?**
Feature

**What does this PR do? Why is it needed?**
Adjust incorrect RPC HTTP responses: return 404 or 400 instead of 500
**Which issues(s) does this PR fix?**

Fixes #15546 

**Other notes for review**

- I updated the status code to 400 or 404 when the wrapped error indicates that the database was not found
- All other responses still return 500
- If I missed any error wrapping cases or misunderstood the issue, please let me know.

**Acknowledgements**

- [x] I have read [CONTRIBUTING.md](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md).
- [x] I have included a uniquely named [changelog fragment file](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md#maintaining-changelogmd).
- [x] I have added a description to this PR with sufficient context for reviewers to understand this PR.

Muzry committed:
Update endpoint to return 404 after isOptimistic check

james-prysm commented:
oops i didn't realize you created a pr for this too will review later. I put my PR in draft for now so that I can review yours

james-prysm committed:
Merge branch 'develop' into update-not-found-status

james-prysm commented on beacon-chain/rpc/eth/shared/errors.go:
```diff
@@ -44,3 +45,26 @@ func WriteBlockFetchError(w http.ResponseWriter, blk interfaces.ReadOnlySignedBe
+
+	// Check for various database lookup errors that should return 404
+	errMsg := err.Error()
+	if strings.Contains(errMsg, "could not fetch state") ||
```
I think this way of checking is a bit fragile, you can look at how I implemented in https://github.com/OffchainLabs/prysm/pull/15563 I think we should match the lookup errors and be able to handle this in the future if we add any new errors

Muzry replied:
I appreciate your feedback; I’ll revise my PR based on your implementation in #15563.

james-prysm reviewed (changes requested):
I think it's on the right track but I left some additional comments

james-prysm commented on changelog/muzry_update_not_found_status.md:
```diff
@@ -0,0 +1,2 @@
+### Updated
```
I think this is not a supported header

Muzry committed:
Merge branch 'develop' into update-not-found-status

Muzry commented:
> I think it's on the right track but I left some additional comments

Thanks again! I’ve updated the code and added some new wrapped errors, but I realized that I basically copied your implementation from PR #15563. Do you think I should keep working on this or just close the PR?

Muzry committed:
Fix error handling by using predefined errors

james-prysm commented:
> > I think it's on the right track but I left some additional comments
> 
> Thanks again! I’ve updated the code and added some new wrapped errors, but I realized that I basically copied your implementation from PR #15563. Do you think I should keep working on this or just close the PR?

no that's fine I encourage external contributions I wrote mine without seeing that work started on it, thanks

james-prysm committed:
Merge branch 'develop' into update-not-found-status

james-prysm reviewed (dismissed):
will also get radek to review since it's partially using my solution

james-prysm commented:
spoke too soon looks like there are linting issues , you can run `bazel run//:gazelle -- fix` typically for linting issues

Muzry committed:
Merge branch 'develop' into update-not-found-status

Muzry committed:
fix: helpers build.bazel

Muzry commented:
> spoke too soon looks like there are linting issues , you can run `bazel run//:gazelle -- fix` typically for linting issues

Thanks!  I've fixed the linting issues

james-prysm reviewed (dismissed):
waiting on radek for a second pair of eyes

rkapka commented on beacon-chain/rpc/lookup/stater.go:
```diff
@@ -21,6 +21,43 @@ import (
 	"github.com/pkg/errors"
 )
 
+type StateIdDecodeError struct {
```
What is the point of this struct? It doesn't play any role in returning 404. Also it looks like a duplicate of `StateIdParseError` to me

Muzry replied:
You’re right — it’s essentially a duplicate of StateIdParseError. I was initially torn between reusing the existing error and defining a new StateIdDecodeError, as their meanings differ only slightly. I’ll go with StateIdParseError instead.

rkapka commented on beacon-chain/rpc/eth/helpers/sync.go:
```diff
@@ -56,7 +56,8 @@ func IsOptimistic(
 			id, err := hexutil.Decode(stateIdString)
 			if err != nil {
-				return false, err
+				e := lookup.NewStateIdDecodeError(err)
```
Can you use `StateIdParseError`?

Muzry committed:
Merge branch 'develop' into update-not-found-status

Muzry committed:
remove the StateIdDecodeError

Muzry committed:
Merge branch 'develop' into update-not-found-status

Muzry committed:
Merge branch 'develop' into update-not-found-status

james-prysm committed:
Merge branch 'develop' into update-not-found-status

james-prysm commented:
looks good to me will ping radek again

james-prysm commented:
license/claExpected — Waiting for status to be reported stuck on this might need to wait for another merge or some empty commit

Muzry committed:
Merge branch 'develop' into update-not-found-status

Muzry commented:
> license/claExpected — Waiting for status to be reported stuck on this might need to wait for another merge or some empty commit

I noticed that the license/cla status is still pending, so I updated the branch to retry the CI.
